libmultilabel/nn/metrics.py

```diff
@@ -60,18 +62,37 @@ class NDCG(Metric):
     def __init__(self, top_k):
         super().__init__()
         self.top_k = top_k
-        self.add_state("ndcg", default=[], dist_reduce_fx="cat")
+        self.add_state("score", default=torch.tensor(0.0, dtype=torch.float64), dist_reduce_fx="sum")
+        self.add_state("num_sample", default=torch.tensor(0, dtype=torch.int64), dist_reduce_fx="sum")
+        self.add_state("discount", default=1.0 / torch.log2(torch.arange(self.top_k) + 2.0))
 
     def update(self, preds, target):
         assert preds.shape == target.shape
-        # implement batch-wise calculations instead of storing results of all batches
-        self.ndcg += [self._metric(p, t) for p, t in zip(preds, target)]
+        dcg = self._dcg(preds, target)
+        idcg = self._idcg(target)
+        ndcg = torch.nan_to_num(dcg / idcg, nan=0.0)
+        self.score += ndcg.sum()
+        self.num_sample += preds.shape[0]
 
     def compute(self):
-        return torch.stack(self.ndcg).mean()
-
-    def _metric(self, preds, target):
-        return retrieval_normalized_dcg(preds, target, k=self.top_k)
+        score = self.score / self.num_sample
+        return score
+
+    def _dcg(self, preds, target):
+        _, sorted_top_k_idx = torch.topk(preds, k=self.top_k)
+        gains = target.take_along_dim(sorted_top_k_idx, 1)
+        dcg = (gains * self.discount).sum(1)
```

Comment:
Please use `torch.dot`.

Reply:
Please refer to https://discuss.pytorch.org/t/dot-product-batch-wise/9746/11. (a * b).sum(dim=1) has the best performance regarding batch dot product.